src/ipld/frozen_cids.rs

```diff
@@ -0,0 +1,176 @@
+///
+/// We use `Box<[...]>` to save memory, avoiding vector overallocation.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FrozenCids(Box<[CidVariant]>);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+pub struct IntoIter<'a> {
+    cids: std::slice::Iter<'a, CidVariant>,
+}
+
+impl<'a> IntoIterator for &'a FrozenCids {
+    type Item = Cid;
+    type IntoIter = IntoIter<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter {
+            cids: self.0.iter(),
+        }
+    }
+}
+
+impl<'a> Iterator for IntoIter<'a> {
+    type Item = Cid;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.cids.next() {
+            Some(cid) => match cid {
+                CidVariant::V1DagCborBlake2b(bytes) => {
+                    Some(Cid::from(CidVariant::V1DagCborBlake2b(*bytes)))
+                }
+                CidVariant::Generic(cid) => Some(*cid.clone()),
+            },
+            None => None,
+        }
+    }
+}
+
+impl Serialize for FrozenCids {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        Vec::<Cid>::from(self).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for FrozenCids {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
+
+impl FromIterator<Cid> for FrozenCids {
+    fn from_iter<T: IntoIterator<Item = Cid>>(iter: T) -> Self {
+        FrozenCids::from(iter.into_iter().collect::<Vec<_>>())
+    }
+}
+
+impl From<Vec<Cid>> for FrozenCids {
+    fn from(cids: Vec<Cid>) -> Self {
+        let mut small_cids = Vec::with_capacity(cids.len());
+        for cid in cids {
+            match cid.into() {
+                CidVariant::V1DagCborBlake2b(bytes) => {
+                    small_cids.push(CidVariant::V1DagCborBlake2b(bytes))
+                }
+                _ => small_cids.push(CidVariant::Generic(Box::new(cid))),
+            }
+        }
+        FrozenCids(small_cids.into_boxed_slice())
+    }
+}
+
+impl From<FrozenCids> for Vec<Cid> {
+    fn from(frozen_cids: FrozenCids) -> Self {
+        Vec::<Cid>::from(&frozen_cids)
+    }
+}
+
+impl From<&FrozenCids> for Vec<Cid> {
+    fn from(frozen_cids: &FrozenCids) -> Self {
+        let mut cids = Vec::with_capacity(frozen_cids.0.len());
+        for cid in frozen_cids.into_iter() {
+            match cid.into() {
+                CidVariant::V1DagCborBlake2b(bytes) => {
+                    cids.push(Cid::from(CidVariant::V1DagCborBlake2b(bytes)))
+                }
+                _ => cids.push(cid),
+            }
+        }
+        cids
+    }
+}
+
+impl FrozenCids {
+    pub fn push(&self, cid: Cid) -> Self {
+        let mut cids = Vec::<Cid>::from(self);
+        cids.push(cid);
+        FrozenCids::from(cids)
+    }
```

Comment:
I don't think a `FrozenCids` struct should have a `push` method

Reply:
Good point, this was a leftover artifact from the previous iterations of this PR. Removed.